components/feature/pwa/src/main/java/mozilla/components/feature/pwa/WebAppInterceptor.kt

```diff
@@ -0,0 +1,55 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.pwa
+
+import android.content.Context
+import android.content.Intent
+import android.net.Uri
+import kotlinx.coroutines.runBlocking
+import mozilla.components.concept.engine.EngineSession
+import mozilla.components.concept.engine.request.RequestInterceptor
+import mozilla.components.feature.intent.ext.putUrlOverride
+import mozilla.components.feature.pwa.intent.WebAppIntentProcessor
+
+/**
+ * This feature implements use cases for detecting and handling redirects to standalone web app activities.
+ *
+ * In the case of Android Intents that are not installed, and with no fallback, the user is prompted
+ * to search the installed market place.
+ *
+ * @param shortcutManager current shortcut manager instance to lookup web app install states
+ */
+@Suppress("LongParameterList")
+class WebAppInterceptor(
+    private val manifestStorage: ManifestStorage
+) : RequestInterceptor {
+
+    override fun onLoadRequest(
+        engineSession: EngineSession,
+        uri: String,
+        hasUserGesture: Boolean,
+        isSameDomain: Boolean
+    ): RequestInterceptor.InterceptionResponse? {
+        return runBlocking {
```

Comment:
I'm wary about having blocking IO on every request, but `AppLinksInterceptor` is similar. @rocketsroger @mcomella can you take a look?

Reply:
I also had doubts about this when I wrote it. It doesn't feel any different, but I can not say what the real performance impact is.

Reply:
Thanks for looping me in - sorry for the delay, I'm behind on my email.

I haven't measured it but we shouldn't block page load requests on IO. In general, we should never be doing blocking IO on any critical code paths (such as startup and page load) or on the main thread (I'm guessing this isn't the main thread but I wanted to include it for completeness): `runBlocking` is a red flag that we should be very careful about adding.

Some possible solutions:
- if possible, cache the results so we don't need to read from disk (seems based on the 
- kick off an asynchronous query for this value. When it returns, if it returns true, cancel the load and launch the intent (if I'm reading this correctly)
  - I'm slightly concerned that reading from this disk during page load could affect page load speed... but I'm not sure what to do about that @acreskeyMoz Do you have thoughts on this?

Reply:
@NotWoods caching all scopes would be the easiest solution. We could just load all installed scopes into a cache. @mcomella when should we populate the cache though? Can we do this during start up, or would it negatively affect the app launch? We also would have to regularly refresh the cache.

`onLoadRequest` acutally expects an async value so doing this async would also be possible, but since android-components implemented the `RequestInterceptor`s synchronous, this would mean we have to now make interceptors optionally asynchronous. That seems a bit complicated.

Reply:
Can you please add an example on how you will be using this in the sample browser?

Reply:
> @mcomella when should we populate the cache though? Can we do this during start up, or would it negatively affect the app launch? We also would have to regularly refresh the cache.

We have a visual completeness queue that can run code just after startup: https://searchfox.org/mozilla-mobile/rev/d81a3c3fce58d3a6c3ebd3257ad1566c9bd462ef/fenix/app/src/main/java/org/mozilla/fenix/FenixApplication.kt#192 Be sure to launch it on a background thread though, assuming it doesn't need to be on the main thread.

> We also would have to regularly refresh the cache.

I don't know how Fenix does periodic jobs like this – @NotWoods do you know?

Reply:
Rather than being periodic, we could just update the cache whenever a new PWA is installed. Old PWAs just "expire" after a certain amount of days so I think just re-populating the cache on startup is fine.

Reply:
> Can you please add an example on how you will be using this in the sample browser? I am a bit worried about adding more `RequestInterceptor` since every `onLoadRequest` goes though them.

@rocketsroger I added this to example browser now. While doing so I noticed that it currently depends on the `AppLinkFeature` to actually launch the `Intent` because `AppLinkFeature` does exactly the right thing here. Is this acceptable or do we need a separate implementation of `onLaunchIntentRequest`?

Reply:
>       * I'm slightly concerned that reading from this disk during page load could affect page load speed... but I'm not sure what to do about that @acreskeyMoz Do you have thoughts on this?

Yes, any delay (even 10-20ms) in the initial pageload request will end up hurting our visual metrics. (Especially for simple and fast pages like www.google.com)
I can help to evaluate this change when it's ready.

Reply:
@NotWoods @mcomella @acreskeyMoz this is no longer blocking and uses a cache which is populated during start up.